### PR TITLE
[Rust] RcU32

### PR DESCRIPTION
**Pull request by Nima-Rahimi-Foroushaani**
It has not been cleaned up, so some debug prints and notes are in the files. I added my concerns about the proof to `rc_u32.rs` as `TODO` notes.

### CODE REVIEW

File: bin/rust/std/lib.rsspec [hunk 1]
```diff
@@ -26,6 +26,51 @@ mod ptr {
...
+        //@ req true;
+        //@ ens *Box_ptr(result) |-> x;
+        // req thread_token(?t) &*& <T>.own(t, x);
+        // ens thread_token(t) &*& Box_own::<T>(t, Box_ptr(result));
+    }
+    impl<T, A> Box<T, A> {
+        fn leak<'a>(x: Box<T, std::alloc::Global>) -> &'a mut T;
+        //@ req thread_token(?t) &*& Box_own::<T>(t, Box_ptr(x));
+        //@ ens thread_token(t) &*& (<T>.full_borrow_content(t, result))() &*& alloc_block(result, std::mem::size_of::<T>()) &*& result == Box_ptr(x);
+    }
```
Comment (btj) [inline bin/rust/std/lib.rsspec:73]: Since Box is not being used, leave it out of this PR.

File: tests/rust/safe_abstraction/rc_u32.rs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        }
+    }
+
+    // The original version in std lib is not an `unsafe` tagged method
+    unsafe fn inner<'a>(&'a self) -> &'a RcBoxU32
+    //@ req [?q](*self).ptr |-> ?nnp;
+    //@ ens [q](*self).ptr |-> nnp &*& result == std::ptr::NonNull_ptr(nnp);
+    {
+        self.ptr.as_ref()
+    }
```
Comment (btj) [inline tests/rust/safe_abstraction/rc_u32.rs:159]: Just inline?

File: tests/rust/safe_abstraction/rc_u32.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+/*@
+// TODO: Prove or make VeriFast to do it automatically by `close`
+lem close_dlft_pred_(dk: lifetime_t, gid: usize) -> real;
+    req [?q]lifetime_token(dk) &*& [?q1]ghost_cell(gid, false);
+    ens [result]dlft_pred(dk)(gid, false) &*& [q - result]lifetime_token(dk) &*& [q1 - result]ghost_cell(gid, false) &*&
+        result < q &*& result < q1;
+//{
+//    if q < q1 { return q/2; } else { return q1/2; }
```
Comment (btj) [inline tests/rust/safe_abstraction/rc_u32.rs:189]: Insert the following:
```
    lifetime_token_inv(dk);
    ghost_cell_fraction_info(gid);
```

**Review by btj — CHANGES_REQUESTED** [review]
Excellent! If you just apply the few requested changes and do a bit of cleanup, we can merge this!

File: bin/rust/rust_belt/lifetime_logic.gh [hunk 4]
```diff
@@ -277,8 +278,13 @@ lemma void close_frac_borrow_atomic(predicate(;) P, lifetime_t k);
...
 // frac_bor_iff in RustBelt repo
 lemma void frac_borrow_implies(lifetime_t k, predicate(;) P, predicate(;) P1);
-    requires frac_borrow(k, P) &*& is_implies(?f, P, P1) &*& is_implies(?f1, P1, P);
-    ensures frac_borrow(k, P1) &*& is_implies(f, P, P1) &*& is_implies(f1, P1, P);
+    requires [_]frac_borrow(k, P) &*& is_implies(?f, P, P1) &*& is_implies(?f1, P1, P);
+    ensures [_]frac_borrow(k, P1) &*& is_implies(f, P, P1) &*& is_implies(f1, P1, P);
+
+// LftL-incl-fract
+lemma void frac_borrow_lft_incl(lifetime_t k, lifetime_t k1);
+    requires [_]exists(?q) &*& [_]frac_borrow(k, lifetime_token_(q, k1));
```
[the remark this answers is not in the document]
Comment (btj) [reply bin/rust/rust_belt/lifetime_logic.gh:286]: You seem to be arguing for the use of `[_]exists` in predicate definitions. That seems different from the use of `[_]exists` in API preconditions? Here, it seems better if the lemma takes `q` as a parameter?

[the remark this answers is not in the document]
Comment (Nima-Rahimi-Foroushaani) [reply bin/rust/rust_belt/lifetime_logic.gh:286]: You are right! Somehow I thought you were talking about `[_]exists` in the `Rc` proof.

**Review by btj — APPROVED** [review]

**Comment by btj** [comment]
🎉

**Pull request merged**